### ISSUE TITLE
Since 2.0.0 smartcard is not working anymore

### ISSUE THREAD

**Issue by marlemion** [FreeRDP/FreeRDP#6068]
I am on arch connecting to a Windows 10 machine at work (obviously - corona).

We need to have a smartcard working to connect to a specific service from this machine. The smartcard is connected locally to the arch client and was redirected perfectly before arch switched to 2.0.0 a few days back. Before, there was 2.0.0-rc4.

As I do not know what kind of debug information you need, I will wait until I get advise from your side. down- and upgrading on arch is easy, even package creation from git is easy, so debugging should not be a problem. Thanks! (For now I reverted back to 2.0.0-rc4 and it is working again.

**Comment by akallabeth** [comment]
We implemented a newer smartcard channel version.
That contains a 'hack' server side to detect which behaviour to use.
Try `/client-build-number:2600` or similar to force old behaviour. See [[MS-RDPESC] 1.7 Versioning and Capability Negotiation](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpesc/0428ca28-b4dc-46a3-97c3-01887fa44a90)

**Comment by akruis** [comment]
I have a similar problem and was able to bisect it: since commit c6e675bfb789dcb5e63788d9588f32211fe6fc14 (Always copy null terminator when filtering smartcard list) my smartcard reader is no longer accessible from the remote side.

I call xfreerdp like this:
`xfreerdp '/smartcard:SCM Microsystems Inc. SCR 331 [CCID Interface] (60102A8A) 00 00' /client-build-number:2600`

**Comment by akallabeth** [comment]
@akruis does it work by simply using `/smartcard` ?

**Comment by akruis** [comment]
> @akruis does it work by simply using `/smartcard` ?

No, it does not work. Omitting the name makes no difference.

**Comment by frispete** [comment]
The new smartcard channel handling broke Kobil mIdentity tokens, as being used from Datev here in Germany. 

I build FreeRDP for openSUSE on a regular base [here](https://build.opensuse.org/package/show/home:frispete:RemoteDesktop/freerdp). Up to some build in March, all was well, but a few days ago, after switching to post 2.0 FreeRDP builds, the mIdentity fails. `/client-build-number:2600` restores it again, but the log is spammed with:
```
[14:01:10:478] [22702:24549] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:10:792] [22702:24543] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeW failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:11:648] [22702:24552] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:12:835] [22702:24556] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:14:005] [22702:24558] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:14:819] [22702:24554] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeW failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:15:190] [22702:24560] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:16:376] [22702:24568] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:17:560] [22702:24570] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:18:732] [22702:24572] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:18:842] [22702:24563] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeW failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:19:917] [22702:24574] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:21:103] [22702:24578] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:22:288] [22702:24588] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:22:867] [22702:24576] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeW failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:23:475] [22702:24590] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:24:659] [22702:24594] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:25:845] [22702:24596] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:26:891] [22702:24592] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeW failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:27:016] [22702:24603] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:28:185] [22702:24607] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:29:355] [22702:24609] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:30:526] [22702:24611] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:30:916] [22702:24605] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeW failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:31:696] [22702:24615] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:32:881] [22702:24631] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:34:067] [22702:24633] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:34:926] [22702:24618] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeW failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:35:237] [22702:24635] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:36:408] [22702:24640] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:37:593] [22702:24647] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:38:781] [22702:24649] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:39:951] [22702:24637] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeW failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:40:948] [22702:24652] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
[14:01:41:119] [22702:24657] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetStatusChangeA failed with error SCARD_E_TIMEOUT [-2146435062]
```

Let me know, if I can provide more data points.

**Comment by akallabeth** [comment]
@frispete can you try a build from master? There was #6098, which might fix the behaviour for newer client build numbers.

**Comment by frispete** [comment]
@akallabeth Forgot to mention, that my build contains #6098 already, using 5b98aa751 as of 20200504T120939.

**Comment by akallabeth** [comment]
@frispete ok, so considering the spamned log as the only remaining issue?
The 'which smartcard works with which protocol version' game you'll have to play, sorry. (Others need it to be able to login, ... )
Not really nice but not much we can do to detect such stuff.

**Comment by frispete** [comment]
> @frispete ok, so considering the spamned log as the only remaining issue?

@akallabeth Any ide, how to deal with this. My RDF session usually runs 24/7 due to some Datev drain bamage. One error log per second isn't that nice (I know how to suppress them, but that's not the issue here).

> The 'which smartcard works with which protocol version' game you'll have to play, sorry. (Others need it to be able to login, ... )
> Not really nice but not much we can do to detect such stuff.

Sure, understood. What's a little disturbing is: rc4 and before worked rock solid (since ages, btw), while the 2.0 release and later doesn't work at all in this respect.  

Even more interesting would be to locate the issues, why the new channel code doesn't work. 
If I can provide anything to further track down this issue, let me know.

**Comment by akallabeth** [comment]
@frispete some functions in newer channel are not supported by pcsc-lite or the devices as far as I can tell.
Not sure how to address that.
Rc4 and older did use the older version only, so no surprise there.

**Comment by frispete** [comment]
@akallabeth to be honest, I didn't expect freerdp the culprit here in the first place. Therefor, I switched to git snapshots of [libusb-1_0](https://build.opensuse.org/package/show/home:frispete:Tumbleweed/libusb-1_0), [pcsc-lite](https://build.opensuse.org/package/show/home:frispete:Tumbleweed/pcsc-lite), and [pcsc-ccid](https://build.opensuse.org/package/show/home:frispete:Tumbleweed/pcsc-ccid), before I started to suspect FreeRDP. I mention this to give you an idea of how much trust I'm placing in your project. :wink: 

I can exclude a deficit on the device side as well, since I tested a couple of them. If these devices wouldn't be supported well in a pure windows environment (w2k-something server, w10pro clients), about 2/3's of **all** German tax consultants would be unemployed, not to mention a significant part of their clients with online accounts, that wouldn't be able to process any tax related work... (just to give you an idea on the ubiquity of this device).

As a starter, I've submitted #6145 now.

**Comment by akallabeth** [comment]
@frispete thank you for the trust and sorry, maybe I did not express the issue well enough.

What I meant is that some part of the windows `PCSC` API is not available with `PCSC-lite` and we can not redirect that properly, therefore some smartcards will not work with the newer version of the redirection channel. (if they use any of the functions we can not redirect)

When using the older version, only a subset of the API is used and many more devices work (as `PCSC-lite` does cover the API used) during a session, but in some scenarios (logon, ...) these might fail and only be successful with the newer channel version.

So, as I´m not really interested in writing hardware drivers for every smartcard device out there, I can´t do much to improve the situation from the FreeRDP side.

**Comment by frispete** [comment]
@akallabeth Thank you for elaboration.

> So, as I´m not really interested in writing hardware drivers for every smartcard device out there, I can´t do much to improve the situation from the FreeRDP side.

This is more than understandable

With #6145 applied, things are fine for me again. Thanks a lot for everything and sorry for being so sticky...

**Comment by akallabeth** [comment]
Fixed with #6152 and #6151
